### ISSUE TITLE
How to create a Tensorflow Dataset without labels? Input 'filename' of 'ReadFile' Op has type float32 that does not match expected type of string

### ISSUE THREAD

**Issue by xenon3dfx** [tensorflow/tensorflow#45397]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
-Linux Ubuntu 20.04
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.3.1
- Python 3.8.5
- CUDA/cuDNN version: 11.0
- GPU model and memory: RTX 2060 8GB

I was trying fit my model using instances of the tf.data.Dataset class. I need the **repeat()** method to perform a custom image augmentation for every epoch.
 
Using Tensorflow 2.3.1, I'm trying to create an instance of the class **tf.data.Dataset** without labels, from the images I have stored in  **.png** files in a folder './Folder/'. For creating the minimal working sample, I think the only relevant line is the one where I am calling **tf.keras.preprocessing.image_dataset_from_directory**. 
The class definition is [here](https://github.com/tensorflow/tensorflow/blob/v2.3.0/tensorflow/python/keras/preprocessing/image_dataset.py#L34-L206):  

**Standalone code to reproduce the issue**

After storing some PNG files in the folder ./Folder/' , the minimal working sample is just this line:
   ` dataset = tf.keras.preprocessing.image_dataset_from_directory('./Folder/',label_mode=None,batch_size=100)`

I expected the tf.data.Dataset instance _dataset_ to be created, and a message:
 Found 449 files ...

Instead, I get the message:
`Found 0 files belonging to 0 classes.
`, followed by this error:
```
Traceback (most recent call last):
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/framework/op_def_library.py", line 465, in _apply_op_helper
    values = ops.convert_to_tensor(
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/framework/ops.py", line 1473, in convert_to_tensor
    raise ValueError(
ValueError: Tensor conversion requested dtype string for Tensor with dtype float32: <tf.Tensor 'args_0:0' shape=() dtype=float32>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "04-vaeAnomalyScores.py", line 135, in <module>
    historicKLD, encoder, decoder, vae = artVAE_Instance.run_autoencoder()  # Train
  File "/media/roi/9b168630-3b62-4215-bb7d-fed9ba179dc7/images/largePatches/artvae.py", line 403, in run_autoencoder
    trainingDataSet = self.loadImages(self.trainingDir)
  File "/media/roi/9b168630-3b62-4215-bb7d-fed9ba179dc7/images/largePatches/artvae.py", line 230, in loadImages
    dataset,paths = tf.keras.preprocessing.image_dataset_from_directory(dir[:-1]+'Downscaled',label_mode=None,batch_size=self.BATCH_SIZE, 
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/keras/preprocessing/image_dataset.py", line 192, in image_dataset_from_directory
    dataset = paths_and_labels_to_dataset(
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/keras/preprocessing/image_dataset.py", line 219, in paths_and_labels_to_dataset
    img_ds = path_ds.map(
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 1695, in map
    return MapDataset(self, map_func, preserve_cardinality=True)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 4041, in __init__
    self._map_func = StructuredFunctionWrapper(
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 3371, in __init__
    self._function = wrapper_fn.get_concrete_function()
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 2938, in get_concrete_function
    graph_function = self._get_concrete_function_garbage_collected(
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 2906, in _get_concrete_function_garbage_collected
    graph_function, args, kwargs = self._maybe_define_function(args, kwargs)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 3213, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 3065, in _create_graph_function
    func_graph_module.func_graph_from_py_func(
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py", line 986, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 3364, in wrapper_fn
    ret = _wrapper_helper(*args)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 3299, in _wrapper_helper
    ret = autograph.tf_convert(func, ag_ctx)(*nested_args)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py", line 255, in wrapper
    return converted_call(f, args, kwargs, options=options)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py", line 532, in converted_call
    return _call_unconverted(f, args, kwargs, options)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/autograph/impl/api.py", line 339, in _call_unconverted
    return f(*args, **kwargs)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/keras/preprocessing/image_dataset.py", line 220, in <lambda>
    lambda x: path_to_image(x, image_size, num_channels, interpolation))
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/keras/preprocessing/image_dataset.py", line 228, in path_to_image
    img = io_ops.read_file(path)
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/ops/gen_io_ops.py", line 574, in read_file
    _, _, _op, _outputs = _op_def_library._apply_op_helper(
  File "/home/roi/.local/lib/python3.8/site-packages/tensorflow/python/framework/op_def_library.py", line 492, in _apply_op_helper
    raise TypeError("%s expected type of %s." %
TypeError: Input 'filename' of 'ReadFile' Op has type float32 that does not match expected type of string.
```

**Comment by Saduf2019** [comment]
@xenon3dfx 
Please share simple indented stand alone code to replicate the issue or if possible share a colab gist with the error.

**Comment by xenon3dfx** [comment]
```
import tensorflow as tf
dataset = tf.keras.preprocessing.image_dataset_from_directory('./folder/',label_mode=None,batch_size=100)
```

Just put png files in the ./folder/ directory

**Comment by aginpatrick** [comment]
Have the same issue. Did you manage to solve the problem @xenon3dfx ?

**Comment by patfla** [comment]
By using pdb to step through code in image_dataset_from_directory and underneath it, I realized that filenames are very tied to classes (as they probably should be) and that the relevant message is not "Op has type float32 that does not match" but rather "Found 0 files belonging to 0 classes."  The first is, if you will, an artifact of the second.

I have (to test) three images in three files: a.jpg, b.jpg, and c.jpg - all originally in one subdirectory, picts.  Classes are also tightly tied to (further sub) directories.  My solution was to create subdirectories A, B and C and place the files appropriately.  Also I needed to not take the default value for class_names and my code reads as follows:

`resultsa=tf.keras.preprocessing.image_dataset_from_directory("d:\\ML\\test\\picts",label_mode=None,class_names=["A","B","C"])`

As best I understand (I didn't want to spend forever stuck on this) your directories probably need to have the same names as your classes.
